kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/records/BatchAwareMemoryRecordsBuilder.java

```diff
@@ -0,0 +1,236 @@
+                                                            boolean isControlBatch,
+                                                            int partitionLeaderEpoch,
+                                                            long deleteHorizonMs) {
+        maybeAppendCurrentBatch();
+        // MRB respects the initial position() of buffer, so this doesn't overwrite anything already in buffer
+        builder = new MemoryRecordsBuilder(buffer,
+                magic,
+                compressionType,
+                timestampType,
+                baseOffset,
+                logAppendTime,
+                producerId,
+                producerEpoch,
+                baseSequence,
+                isTransactional,
+                isControlBatch,
+                partitionLeaderEpoch,
+                0, // TODO think about limiting the size that the buffer can grow to
+                deleteHorizonMs);
+        return this;
+    }
+
+    public BatchAwareMemoryRecordsBuilder addBatch(CompressionType compressionType,
+                                                   TimestampType timestampType,
+                                                   long baseOffset) {
+        return addBatch(RecordBatch.CURRENT_MAGIC_VALUE,
+                compressionType,
+                timestampType,
+                baseOffset,
+                timestampType == TimestampType.LOG_APPEND_TIME ? System.currentTimeMillis() : RecordBatch.NO_TIMESTAMP,
+                RecordBatch.NO_PRODUCER_ID,
+                RecordBatch.NO_PRODUCER_EPOCH,
+                RecordBatch.NO_SEQUENCE,
+                false,
+                false,
+                RecordBatch.NO_PARTITION_LEADER_EPOCH,
+                RecordBatch.NO_TIMESTAMP);
+    }
+
+    /**
+     * Starts a batch, with batch parameters taken from the given {@code templateBatch}.
+     * @param templateBatch The batch to use as a source of batch parameters
+     * @return this builder
+     */
+    public @NonNull BatchAwareMemoryRecordsBuilder addBatchLike(RecordBatch templateBatch) {
+        return addBatch(templateBatch.magic(),
+                templateBatch.compressionType(),
+                templateBatch.timestampType(),
+                templateBatch.baseOffset(),
+                0,
+                templateBatch.producerId(),
+                templateBatch.producerEpoch(),
+                templateBatch.baseSequence(),
+                templateBatch.isTransactional(),
+                templateBatch.isControlBatch(),
+                templateBatch.partitionLeaderEpoch(),
+                templateBatch.deleteHorizonMs().orElse(RecordBatch.NO_TIMESTAMP));
+    }
+
+    private void maybeAppendCurrentBatch() {
+        if (haveBatch()) {
+            appendCurrentBatch();
+        }
+    }
+
+    private void appendCurrentBatch() {
+        // Calling build will write out the (possibly compressed) batch bytes into this.buffer
+        builder.build();
+    }
```

Comment:
I'm suggesting this defensively, to help catch the case where we erroneously try to append to a built batch.
```suggestion
        builder = null;
    }
```

Reply:
I ended up using the `builder.isClosed` for the state mangement, rather than adding a field on BAMRB, which means I can't nullify the builder.